### ISSUE TITLE
How to kill GPU lock on actor?

### ISSUE THREAD

**Issue by abrahamrhoffman** [ray-project/ray#1229]
### System information
- Linux Ubuntu 16.04
- Ray installed from binary
- Ray Version: 0.2.2
- Python Version: 3.5.2

### Describe the problem
Need a method for killing GPU lock for an actor.

### Demonstrate the problem

$ ray start --huge-pages --plasma-directory=/mnt/hugepages --num-gpus=3 --num-cpus=12 --head --redis-port=6379 --num-workers=3

```
@ray.remote(num_gpus=1)
class GPU(object):
    def __init__(self):
        pass

GPU.remote() # Works 1st Time
# <---- Kill GPU lock here
GPU.remote() # Fails 2nd Time
```
```
Exception: Could not find a node with enough GPUs or other resources to create this actor. The local scheduler information is [ {'Deleted': False, 'DBClientID': 'c22dd34381525f6d18edf55f3cd3ec6fe661b548', 'ClientType': 'local_scheduler', 'NumGPUs': 3.0, 'NumCPUs': 12.0, 'AuxAddress': '192.168.1.200:52507', 'LocalSchedulerSocketName': '/tmp/scheduler7534451'}].
```

**Comment by robertnishihara** [comment]
When the first actor goes out of scope, it should release the GPU resource. For example, doing

```
actor1 = GPU.remote()
del actor1
```

should release the GPU resources. Normally you shouldn't have to call `del`, it should just happen when `actor1` goes out of scope.

Note that the GPU resources are not released synchronously. That is, a "destructor" task actually is scheduled on the actor and runs and releases the resources, so if you do 

```
actor1 = GPU.remote()
del actor1
actor2 = GPU.remote()
```

it probably won't work, but it will work if you do

```
import time

actor1 = GPU.remote()
del actor1
time.sleep(2)
actor2 = GPU.remote()
```

**Comment by abrahamrhoffman** [comment]
@robertnishihara 

This makes sense logically; however from a scheduling perspective, it would be convenient (perhaps essential) to check the status of the lock itself.

### Current
```
import ray, time
@ray.remote(num_gpus=1)
class GPU(object):
    def __init__(self):
        pass

actor1 = GPU.remote()
time.sleep(int)
del actor1
```

### Proposed
```
import ray, time
@ray.remote(num_gpus=1)
class GPU(object):
    def __init__(self):
        pass

actor1 = GPU.remote()
actor1.gpu_status()
actor1.delete()
```

My defense for this request:
1. Ray should be considered a one-stop shop for scheduling
2. User defined lock checks aren't a bad thing, but it isn't obvious from the consumer perspective that ray -can't- naturally provide this feedback.
3. GPU is a special resource, in the fact that it doesn't handle attempted over-saturation like CPU bound ops.
4. In most Machine Learning applications (and at least in mine), I need to have explicit control over every GPU. The idea that I could do this was what initially attracted me to Ray.
5. Lots of modern GPU scheduling libraries provide this feedback.

**Comment by ericl** [comment]
@abrahamrhoffman I think https://github.com/ray-project/ray/issues/1049 can also resolve the issue. After that change, you can wait for resource release by calling `ray.wait()` on a task that shuts down the actor as follows:

```
import ray
@ray.remote(num_gpus=1)
class GPU(object):
    def __init__(self):
        pass

    def kill(self):
       exit(0)


actor1 = GPU.remote()
ray.wait([actor1.kill.remote()])
# resources have been guaranteed to be released
```

**Comment by robertnishihara** [comment]
Thanks @abrahamrhoffman. One change we'll probably make in the future is that attempting to create an actor won't raise an exception when there aren't enough GPUs. Instead, the actor creation event will be queued and will happen when a GPU becomes available. We may want to print some warnings so that people know why the actor hasn't been created yet.

As for checking the lock, what exactly does `actor1.gpu_status()` return?

**Comment by abrahamrhoffman** [comment]
@ericl 

Thank you for the suggestion. I do think ray.wait() is a nice option in some instances:
- When our scientists are developing proofs-of-concepts
- When a small or non-production workload is being computed.

#### Use Case
However, the over-arching issue is that we need to use our production ray cluster to perform ML inference in sub-1000ms. In order to achieve that, we must apply optimization algorithms to our hot and cold storage: effectively pushing numpy arrays into the ray cluster (plasma object store) when inference is predicted. We need to highly tune the cluster, GPUs and associated memories to achieve this.

@robertnishihara 

An automated GPU queue is a nice feature. However, because we place such high priority on reducing latency in the ray cluster, we would need some fine grained feedback around the queue itself (thoughts?)

`actor.gpu_status()` could return:
- GPU's hosting node in the cluster in the format cluster_memberid[0,1,2,3,gpu_idN]
- The make and model of the utilized GPU [NVIDIA Tesla V100, etc]
- The total available GPU RAM and the utilized amount [384/16384]
- (Optional) The CUDA and cuDNN version installed on the cluster_member

**Comment by ericl** [comment]
Ah I see, is the idea that you'd use gpu status to make scheduling decisions, and there are many actor processes sharing a set of GPUs? I can see how Ray's process level locks might be too heavyweight for that use case.

A couple followup questions --
We've found that some frameworks like tensorflow basically never release their GPU resources so you have to terminate the entire process, is that not an issue for your use case?
Would it be useful to be able to allocate fractional GPU resources to processes? This could allow for more fine-grained multiplexing of processes.

I'm also not sure if it makes sense to expose some of the information e.g. GPU type as an after the fact thing on an actor -- wouldn't it be required to create an actor and have it get scheduled on a particular type of GPU (in which case you'd already know the GPU type?) We have some upcoming support for that in the form of scheduler label constraints.

**Comment by abrahamrhoffman** [comment]
@ericl 

`is the idea that you'd use gpu status to make scheduling decisions` : Yes, this is what we would like. 
`there are many actor processes sharing a set of GPUs?` : Yes. We have multiple libraries contending for GPU resources based on their business purpose. Sharding the GPU memory is essential.

#### Our Workflow:
- A numpy array is placed into the ray cluster

- If the inference algorithm determines the numpy array needs to be inferred:
  - Based on the in-the-moment resource availability, we schedule a GPU inference
- If the inference algorithm determines the numpy array needs to be stored:
  - We move the numpy array to the non-production cluster, execute a transform (non-GPU) and store the array in cold storage.

Later, when batch ML training is performed, we pull from the cold store and place into the non-production cluster. This array becomes part of a larger training job depending on the library (TensorFlow, PyTorch, etc) that is running the distributed job through ray.

`tensorflow basically never release[s] ... GPU resources` : Yes, it is an issue:
- With PyTorch, we have to `del cudaObject;torch.cuda.empty_cache()`
- With TensorFlow, we have to `session.close()` or similar

`Would it be useful to be able to allocate fractional GPU resources to processes? This could allow for more fine-grained multiplexing of processes.` : Yep, this would totally rock.

`wouldn't it be required to create an actor and have it get scheduled on a particular type of GPU` : We're growing into AWS and need to spin up and terminate GPU instances based on unanticipated load spikes. While we would know which instances we pre-allocate for this action (and subsequently their GPU specs) it is much more convenient to query the cluster in-the-moment and retrieve granular details about the available cluster's capabilities.

**Comment by ericl** [comment]
Ok, thanks for the clarifications, it makes a lot of sense.

Though Ray doesn't currently provide a fine-grained status API, I think you could meet a lot of the requirements by tracking the scheduling state at the application level, e.g. keeping a registry of created actors, what tasks they're currently executing, etc. You could feed then feed this state to your systems that manage resource allocation. For example, we use a counter to track the resources used by currently active actors in the `ray.tune` app for admission control: https://github.com/ray-project/ray/blob/master/python/ray/tune/trial_runner.py#L200

The only thing that may be a fundamental limitation is the fractional GPUs thing, which prevents actors from sharing GPUs.

**Comment by robertnishihara** [comment]
We currently can't do fractional GPUs (#402), although we'll soon provide more customizable resource specifications, e.g., something like #1236, and that could be used to implement your own custom fractional GPU support (though it would be the applications responsibility actually enforce fractional GPU memory limits).

**Comment by abrahamrhoffman** [comment]
@ericl @robertnishihara 

For me to summarize:

`Fractional GPU`
- Fractional GPU support could be enforced in a custom wrapper
  - Specifically, memory limits, memory usage, GPU models, and CUDA/cuDNN metadata will need to  be managed by by our wrapper.
  - @ericl Thank you for the `counter` reference

`Scheduling State`
- Can be tracked at the application level and fed to consuming systems.

### Path Forward
- #1236 for custom resources (GPU sharding)
- `ray.wait()` for actor deletion and subsequent deletion validation prior to removal
- Track resource consumption at our application level

**Comment by abrahamrhoffman** [comment]
For posterity:

```
@ray.remote
def services():
    return [method for method in dir(ray.services) if callable(getattr(ray.services, method))]

ray.get(services.remote())

['ObjectStoreAddress',
 'OrderedDict',
 'address',
 'address_to_ip',
 'all_processes_alive',
 'cleanup',
 'get_ip_address',
 'get_node_ip_address',
 'get_port',
 'kill_process',
 'namedtuple',
 'new_log_files',
 'new_port',
 'random_name',
 'record_log_files_in_redis',
 'start_global_scheduler',
 'start_local_scheduler',
 'start_log_monitor',
 'start_monitor',
 'start_objstore',
 'start_ray_head',
 'start_ray_node',
 'start_ray_processes',
 'start_redis',
 'start_redis_instance',
 'start_ui',
 'start_worker',
 'try_to_create_directory',
 'wait_for_redis_to_start']
```
